File: loader/src/main/java/de/timolia/lactea/loader/internal/JavassistAnnotations.java
package de.timolia.lactea.loader.internal;

import de.timolia.lactea.loader.module.ModuleDescription.DependencyDescription;
import javassist.bytecode.annotation.Annotation;
import javassist.bytecode.annotation.AnnotationMemberValue;
import javassist.bytecode.annotation.ArrayMemberValue;
import javassist.bytecode.annotation.BooleanMemberValue;
import javassist.bytecode.annotation.MemberValue;
import javassist.bytecode.annotation.StringMemberValue;

/**
 * @author David (_Esel)
 */
public class JavassistAnnotations {
    public static String stringValue(Annotation annotation, String name) {
        StringMemberValue value = (StringMemberValue) annotation.getMemberValue(name);
        return value.getValue();
    }

    public static DependencyDescription[] dependencyDescriptions(Annotation annotation, String name) {
        ArrayMemberValue value = (ArrayMemberValue) annotation.getMemberValue(name);
        MemberValue[] dependencies = value.getValue();
        DependencyDescription[] dependencyDescriptions = new DependencyDescription[dependencies.length];
        for (int i = 0; i < dependencies.length; i++) {
            Annotation dependencyAnnotation = ((AnnotationMemberValue) dependencies[i]).getValue();
            dependencyDescriptions[i] = new DependencyDescription(
                stringValue(dependencyAnnotation, "value"),
                ((BooleanMemberValue) dependencyAnnotation.getMemberValue("required")).getValue()
            );
        }
        return dependencyDescriptions;
    }
}

File: core/src/main/kotlin/de/timolia/lactea/core/module/JavassistAnnotations.kt
package de.timolia.lactea.core.module

import de.timolia.lactea.core.module.ModuleDescription.DependencyDescription
import javassist.bytecode.annotation.*
import javassist.bytecode.annotation.Annotation

internal object JavassistAnnotations {
    fun stringValue(annotation: Annotation, name: String): String {
        val value: StringMemberValue = annotation.getMemberValue(name) as StringMemberValue
        return value.value
    }

    fun dependencyDescriptions(annotation: Annotation, name: String): Array<DependencyDescription> {
        val value: ArrayMemberValue = annotation.getMemberValue(name) as ArrayMemberValue
        return value.value.map {
            val dependencyAnnotation: Annotation = (it as AnnotationMemberValue).value
            DependencyDescription(
                stringValue(dependencyAnnotation, "value"),
                (dependencyAnnotation.getMemberValue("required") as BooleanMemberValue).value
            )
        }.toTypedArray()
    }
}
